src/nvim/decoration.h

```diff
@@ -57,9 +58,28 @@ typedef struct {
   int draw_col;
 } DecorRange;
 
+/// DecorRange can be removed from `DecorState` list in any order,
+/// so we track available slots using a freelist (with `next_free_i`).
+/// The list head is in `DecorState.free_slot_i`.
+typedef union {
+  DecorRange range;
+  int next_free_i;
+} DecorRangeSlot;
+
 typedef struct {
   MarkTreeIter itr[1];
-  kvec_t(DecorRange) active;
+  kvec_t(DecorRangeSlot) slots;
+  kvec_t(int) ranges_i;
+  /// Indices in [0; current_end) of `ranges_i` point to ranges that start
+  /// before current position. Sorted by priority and order of insertion.
+  int current_end;
+  /// Indices in [future_begin, kv_size(ranges_i)) of `ranges_i` point to
+  /// ranges that start after current position. Sorted by starting position.
+  int future_begin;
+  /// Head of DecorRangeSlot freelist. -1 if none are freed.
+  int free_slot_i;
+  /// Index for keeping track of range insertion order.
+  int new_range_ordering;
```

Comment:
Looks like 4+ new fields were added, and a lot of related logic. Is there any way to extract parts of the logic to common functions, to help with understanding it in the future? Or some other way to simplify the code?